odoo/models.py

```diff
@@ -4292,6 +4289,85 @@ def _check_company(self, fnames=None):
+            records.browse().check_access(operation)
+
+        """
+        if not self.env.su and (forbidden := self._check_access(operation)):
+            # best effort: raise the first exception
+            exception = next(iter(forbidden.values()))()
+            # invalidate forbidden records to avoid cache "pollution"
+            if None not in forbidden:
+                self.browse(forbidden).invalidate_recordset()
+            raise exception
+
+    def has_access(self, operation: str) -> bool:
+        """ Return whether the current user is allowed to perform ``operation``
+        on all the records in ``self``. The method is fully consistent with
+        method :meth:`check_access` but returns a boolean instead.
+        """
+        if not self.env.su and (forbidden := self._check_access(operation)):
+            # invalidate forbidden records to avoid cache "pollution"
+            if None not in forbidden:
+                self.browse(forbidden).invalidate_recordset()
+            return False
+        return True
+
+    def _filter_access(self, operation: str):
+        """ Return the subset of ``self`` for which the current user is allowed
+        to perform ``operation``. The method is fully equivalent to::
+
+            self.filtered(lambda record: record.has_access(operation))
+
+        """
+        if not self.env.su and (forbidden := self._check_access(operation)):
+            if None in forbidden:
+                return self.browse()
+            # invalidate forbidden records to avoid cache "pollution"
+            self.browse(forbidden).invalidate_recordset()
+            # records are forbidden, invalidate them to avoid cache "pollution"
+            return self.browse(id_ for id_ in self._ids if id_ not in forbidden)
+        return self
+
+    def _check_access(self, operation: str) -> dict:
+        """ Return a dict mapping forbidden record ids (or ``None``) to a
+        callable that generates an exception. When the current user has no
+        permission to perform ``operation`` on the model in general, the output
+        dict only contains key ``None``. Otherwise, the keys of the dict are
+        the ids of the forbidden records.
+
+        This method provides the base implementation of
+        methods :meth:`check_access`, :meth:`has_access`
+        and :meth:`_filter_access`. The method may be overridden in order to
+        restrict the access to ``self``.
+        """
+        access_model = self.env['ir.model.access']
+        if not access_model.check(self._name, operation, raise_exception=False):
+            return {None: (lambda: access_model._make_access_error(self._name, operation))}
+
+        if any(self._ids):
+            rule_model = self.env['ir.rule']
+            domain = rule_model._compute_domain(self._name, operation) or []
+            allowed = self.sudo().filtered_domain(domain)
+            if len(allowed) < len(self):
```

Comment:
Minor simplification.
```suggestion
            domain = rule_model._compute_domain(self._name, operation)
            if domain and len(allowed := self.sudo().filtered_domain(domain)) < len(self):
```